### PR TITLE
AP_GPS: support GPSx_GNSS_MODE for F9P

### PR DESCRIPTION
**Pull request by andyp1per**
This provides limited support for the GPSx_GNSS_MODE parameter. Specifically it allows you to turn off and on the SBAS. Galileo, Beidou and Glonass constellations. All signals for these constellations are left at the defaults. GPS and QZSS are left enabled as recommended and cannot be changed.

This PR also fixes 2 bugs in the UBLOX driver:
- Extensions are actually 30*N bytes for N extensions. Previously the buffer was sized too small meaning that extensions could be missed.
- 1 byte configuration items only use the LSB, so the other 7 bits need to masked out when checking that a configuration item is enabled or disabled

This work kindly sponsored by LightDynamix

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 1]
```diff
@@ -455,6 +455,68 @@ AP_GPS_UBLOX::_request_next_config(void)
 #endif
         break;
 
+    case STEP_F9: {
+        if (_hardware_generation == UBLOX_F9 && params.gnss_mode != 0) {
+
+            AP_GPS_UBLOX::config_list gnss_cfg[16];
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp:461]: should 16 be UBLOX_MAX_GNSS_CONFIG_BLOCKS?

**Review by tridge — CHANGES_REQUESTED** [review]
i think you'll find that lots of config combinations are not allowed. We need some way to stop this trying forever

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Fixed

**Comment by andyp1per** [comment]
I have verified that all the allowed configurations work.

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 2]
```diff
@@ -455,6 +463,26 @@ AP_GPS_UBLOX::_request_next_config(void)
...
+            uint8_t cfg_count = populate_F9_gnss();
+            // special handling of F9 config
+            if (cfg_count > 0) {
+                CFG_Debug("Sending F9 settings, GNSS=%u", params.gnss_mode);
+                if (!_configure_list_valset(config_GNSS, cfg_count, UBX_VALSET_LAYER_RAM | UBX_VALSET_LAYER_BBR)) {
+                    _next_message--;
+                    break;
+                }
+                // the call will reset the GNSS subsystem so spec says to wait a further 0.5s before sending more messages
+                hal.scheduler->delay(500);
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp:477]: this is unsafe, has to be state machine, this runs in main thread

File: libraries/AP_GPS/AP_GPS_UBLOX.h [hunk 3]
```diff
@@ -518,7 +521,7 @@ class AP_GPS_UBLOX : public AP_GPS_Backend
     struct PACKED ubx_mon_ver {
         char swVersion[30];
         char hwVersion[10];
-        char extension[50]; // extensions are not enabled
+        char extension[30*UBLOX_MAX_EXTENSIONS]; // extensions are not enabled
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.h:524]: need to test this on a ublox m6

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_GPS/AP_GPS_UBLOX.h (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_GPS/AP_GPS_UBLOX.h:524]: Tested. There are no extensions so it works ok

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Fixed

**Comment by MallikarjunSE** [comment]
Hey @andyp1per is it possible to look at F10N too? 
This module doesn't support GLONASS and sometimes very randomly I get GPS still configuring.

**Comment by andyp1per** [comment]
Hi @MallikarjunSE this work was sponsored by LightDyamix and I am done at this point. There are many enhancements that could be considered - including M10 support. Happy to look at this on a commercial basis, but I would need a sample as well to get it working. Cheers Andy

**Comment by tridge** [comment]
@MallikarjunSE I can test the F10N you sent me

**Comment by MallikarjunSE** [comment]
> @MallikarjunSE I can test the F10N you sent me

Thanks @tridge

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 4]
```diff
@@ -273,6 +273,8 @@ const AP_GPS_UBLOX::config_list AP_GPS_UBLOX::config_L5_ovrd_dis[] {
     {ConfigKey::CFG_SIGNAL_L5_HEALTH_OVRD, 0},
 };
 
+AP_GPS_UBLOX::config_list AP_GPS_UBLOX::config_GNSS[UBLOX_MAX_GNSS_CONFIG_BLOCKS*3] {};
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp:276]: certainly should not be initialised, but would be much better to be dynamically allocated as needed, and freed when done, step skipped if we can't allocate the memory

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: Fixed

**Review by tridge — APPROVED** [review]
tested on F10-N and Zed-F9P

**Pull request merged**